sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/RegistrationKeyManager.java

```diff
@@ -0,0 +1,144 @@
+ * from key-store every refresh.
+ *
+ * RegistrationKeyStore is used as object to manage the keys in file/memory.
+ */
+public final class RegistrationKeyManager {
+    private static RegistrationKeyManager registrationKeyManager;
+
+    private KeyGenerator keyGenerator;
+
+    private RegistrationKeyStore registrationKeyStore;
+
+    //The duration we persist keys in key-store
+    public static final int EXPIRATION_TIME_MINUTES = 45;
+
+    private ClientLogger clientLogger = new ClientLogger(RegistrationKeyManager.class);
+
+    private RegistrationKeyManager()  {
+        try {
+            this.keyGenerator = KeyGenerator.getInstance("AES");
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("KeyGenerator failed: " + e.getMessage());
+        }
+        this.keyGenerator.init(256);
+        registrationKeyStore = new RegistrationKeyStore();
+    }
+
+    public static RegistrationKeyManager instance() {
+        if (registrationKeyManager == null) {
+            synchronized (RegistrationKeyManager.class) {
+                if (registrationKeyManager == null) {
+                    registrationKeyManager = new RegistrationKeyManager();
+                }
+            }
+        }
+        return registrationKeyManager;
+    }
+
+    private long getInsertionTime(RegistrationKeyStore.RegistrationKeyEntry entry) {
+        return entry.getCreationTime();
+    }
+
+    private int getNumOfPairs() {
+        int numKeyPairs = 0;
+        try {
+            numKeyPairs = registrationKeyStore.getSize() / 2;
+        } catch (Exception e) {
+            clientLogger.logExceptionAsError(new RuntimeException("Failed to get size from key store", e));
+        }
+        Log.v("RegistrationContainer", "Number of pairs in key-store: " + numKeyPairs);
+        return numKeyPairs;
+    }
+
+    public void refreshCredentials(String path) {
+        synchronized (RegistrationKeyManager.class) {
+            Log.v("RegistrationContainer", "refresh keys");
+            //Fetch latest data from file
+            load(path);
+
+            //Rotate key based on creation time
+            rotateKeysBasedOnTime(path);
+            SecretKey newCryptoKey = keyGenerator.generateKey();
+            String cryptoCredential = secretKeyToStr(newCryptoKey);
+            SecretKey newAuthKey = keyGenerator.generateKey();
+            String authCredential = secretKeyToStr(newAuthKey);
+            long currentTimeMillis = System.currentTimeMillis();
+            RegistrationKeyStore.RegistrationKeyEntry entry = new RegistrationKeyStore.RegistrationKeyEntry(cryptoCredential, authCredential, currentTimeMillis);
+            registrationKeyStore.storeKeyEntry(path, entry);
+        }
+    }
+
+    private void rotateKeysBasedOnTime(String path) {
+        int removedEntry = 0;
+        int size = getNumOfPairs();
+        //Delete expired keys
+        for (int curIndex = 1; curIndex <= size; curIndex++) {
```

Comment:
Why are we starting with 1 and using `<=`, as opposed to starting with 0 and using `<`? I think the error message below where `curIndex` is used could be confusing otherwise.